File: _skels/go-ddd-skel/internal/state/service.go
// Service-layer logic for /api/state.
package state

import (
	"context"
	"errors"
	"fmt"

	"github.com/example/go-ddd-skel/internal/shared"
)

// Service coordinates state.Repository for HTTP routes.
type Service struct {
	repo Repository
}

// NewService builds a state Service.
func NewService(repo Repository) *Service {
	return &Service{repo: repo}
}

// Map returns every key/value pair owned by userID as a flat map.
func (s *Service) Map(ctx context.Context, userID uint) (map[string]string, error) {
	rows, err := s.repo.ListForUser(ctx, userID)
	if err != nil {
		return nil, err
	}
	out := make(map[string]string, len(rows))
	for _, row := range rows {
		out[row.StateKey] = row.StateValue
	}
	return out, nil
}

// Upsert stores value under key for userID, inserting or updating as
// needed.
func (s *Service) Upsert(ctx context.Context, userID uint, key, value string) error {
	if key == "" {
		return fmt.Errorf("%w: state key cannot be empty", shared.ErrValidation)
	}
	row, err := s.repo.GetForUser(ctx, userID, key)
	switch {
	case err == nil:
		row.StateValue = value
		return s.repo.Save(ctx, &row)
	case errors.Is(err, shared.ErrNotFound):
		row = ReactState{
			UserID:     userID,
			StateKey:   key,
			StateValue: value,
		}
		return s.repo.Save(ctx, &row)
	default:
		return err
	}
}

// Delete removes the key for userID. No-ops when the key is absent
// (matches the wrapper-shared 200-with-empty-body contract).
func (s *Service) Delete(ctx context.Context, userID uint, key string) error {
	if key == "" {
		return fmt.Errorf("%w: state key cannot be empty", shared.ErrValidation)
	}
	return s.repo.DeleteForUser(ctx, userID, key)
}

File: _skels/rust-actix-ddd-skel/src/state/service.rs
//! Service-layer logic for `/api/state`.

use std::collections::HashMap;
use std::sync::Arc;

use crate::shared::DomainError;
use crate::state::repository::StateRepository;

#[derive(Clone)]
pub struct StateService {
    repo: Arc<dyn StateRepository>,
}

impl StateService {
    pub fn new(repo: Arc<dyn StateRepository>) -> Self {
        Self { repo }
    }

    /// Return every key/value owned by `user_id` as a flat map.
    pub async fn map(&self, user_id: i32) -> Result<HashMap<String, String>, DomainError> {
        let rows = self.repo.list_for_user(user_id).await?;
        Ok(rows.into_iter().collect())
    }

    pub async fn upsert(
        &self,
        user_id: i32,
        key: &str,
        value: String,
    ) -> Result<(), DomainError> {
        if key.is_empty() {
            return Err(DomainError::Validation(
                "state key cannot be empty".to_string(),
            ));
        }
        self.repo.upsert(user_id, key, value).await
    }

    pub async fn delete(&self, user_id: i32, key: &str) -> Result<(), DomainError> {
        if key.is_empty() {
            return Err(DomainError::Validation(
                "state key cannot be empty".to_string(),
            ));
        }
        self.repo.delete(user_id, key).await
    }
}
